crates/evm/src/execute.rs

```diff
@@ -161,6 +162,152 @@ pub trait BlockExecutorProvider: Send + Sync + Clone + Unpin + 'static {
         DB: Database<Error: Into<ProviderError> + Display>;
 }
 
+/// A factory for creating block execution strategies.
+pub trait BlockExecutionStrategyFactory<DB>
+where
+    DB: Database,
+{
+    /// The specific [`BlockExecutionStrategy`] type this factory produces.
+    type Strategy: BlockExecutionStrategy<DB>;
+
+    /// The error type returned by this factory and its produced strategies.
+    type Error: From<ProviderError>
+        + From<<Self::Strategy as BlockExecutionStrategy<DB>>::Error>
+        + core::error::Error;
+
+    /// Creates a new block execution strategy instance.
+    fn create(
+        &self,
+        block: &BlockWithSenders,
+        total_difficulty: U256,
+    ) -> Result<Self::Strategy, Self::Error>;
+}
+
+/// Defines the strategy for executing a single block.
+pub trait BlockExecutionStrategy<DB> {
+    /// The error type returned by this strategy's methods.
+    type Error: From<ProviderError> + core::error::Error;
+
+    /// Configures the environment before start applying any changes.
+    fn configure_environment(&mut self) -> Result<(), Self::Error>;
+
+    /// Applies any necessary changes before executing the block's transactions.
+    fn apply_pre_execution_changes(&mut self) -> Result<(), Self::Error>;
+
+    /// Executes all transactions in the block.
+    fn execute_transactions(
+        &mut self,
+        block: &BlockWithSenders,
+    ) -> Result<(Vec<Receipt>, u64), Self::Error>;
```

Comment:
for a followup, I think we def want a container type here because returning raw u64 isn't great. that would also make it easier to introduce generics for the receipt which we'll eventually need to do